File: examples/lissajous/lissajous.js
let duration = 5;
let margin = 3;
let pointsNb = 26;
let points = [];

let a = 5;
let b = 2;

function setup(c) {
  c.size(40, 20);
  c.backgroundBg("197");
  c.stroke(".", "255", "197");
}

function draw(c) {
  c.clear();

  let theta = c.loopAngle(duration);
  let x = c.map(Math.sin(a * theta), -1, 1, margin, c.width - margin);
  let y = c.map(Math.sin(b * theta), -1, 1, margin, c.height - margin);
  let x2 = c.map(Math.sin(b * theta), -1, 1, margin, c.width - margin);
  let y2 = c.map(Math.sin(a * theta), -1, 1, margin, c.height - margin);
  if (points.length <= pointsNb * 2) {
    points.push(new Point(x, y, "0"));
    points.push(new Point(x2, y2, "#"));
  }

  for (let i = 0; i < points.length; i++) {
    points[i].update(c);
  }
}

class Point {
  constructor(x, y, char) {
    this.x = x;
    this.y = y;
    this.char = char;
    this.duration = pointsNb;
  }

  update(c) {
    c.strokeText(this.char);
    c.point(this.x, this.y);
    this.duration--;
    if (this.duration < 0) {
      points.shift();
    }
  }
}

function onKey(c, e) {
  if (e.key == "space") {
    c.noiseSeed(Date.now());
  }
}

File: examples/lissajous/lissajous.go
package main

import (
	"context"
	"math"
	"time"

	"github.com/emprcl/runal"
)

const (
	duration = 5
	margin   = 3
	pointsNb = 26
	a        = 5
	b        = 2
)

type point struct {
	x, y     float64
	char     string
	duration int
}

func (p *point) update(c *runal.Canvas) {
	c.StrokeText(p.char)
	c.Point(int(p.x), int(p.y))
	p.duration--
}

var points = []point{}

func main() {
	runal.Run(context.Background(), setup, draw, runal.WithOnKey(onKey))
}

func setup(c *runal.Canvas) {
	c.Size(40, 20)
	c.BackgroundBg("197")
	c.Stroke(".", "255", "197")
}

func draw(c *runal.Canvas) {
	c.Clear()
	theta := c.LoopAngle(duration)
	x := c.Map(math.Sin(a*theta), -1, 1, margin, float64(c.Width-margin))
	y := c.Map(math.Sin(b*theta), -1, 1, margin, float64(c.Height-margin))
	x2 := c.Map(math.Sin(b*theta), -1, 1, margin, float64(c.Width-margin))
	y2 := c.Map(math.Sin(a*theta), -1, 1, margin, float64(c.Height-margin))
	if len(points) <= pointsNb*2 {
		points = append(points, point{x, y, "0", pointsNb})
		points = append(points, point{x2, y2, "#", pointsNb})
	}

	for i := range points {
		points[i].update(c)
	}

	// remove points with a duration of zero
	filtered := points[:0]
	for _, p := range points {
		if p.duration > 0 {
			filtered = append(filtered, p)
		}
	}
	points = filtered
}

func onKey(c *runal.Canvas, e runal.KeyEvent) {
	if e.Key == "space" {
		c.NoiseSeed(time.Now().Unix())
		c.Redraw()
	}
}
